stack_validation.go

```diff
@@ -0,0 +1,44 @@
+package lifecycle
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/buildpacks/imgutil"
+	"github.com/pkg/errors"
+
+	"github.com/buildpacks/lifecycle/cmd"
+	"github.com/buildpacks/lifecycle/platform"
+)
+
+func ValidateStack(stackMD platform.StackMetadata, runImage imgutil.Image) error {
+	buildStackID, err := getBuildStack(stackMD)
+	if err != nil {
+		return err
+	}
+
+	runStackID, err := runImage.Label(platform.StackIDLabel)
+	if err != nil {
+		return errors.Wrap(err, "get run image label")
+	}
+	if runStackID == "" {
+		return errors.New("get run image label: io.buildpacks.stack.id")
+	}
+
+	if buildStackID != runStackID {
+		return errors.New(fmt.Sprintf("incompatible stack: '%s' is not compatible with '%s'", runStackID, buildStackID))
+	}
+	return nil
+}
+
+func getBuildStack(stackMD platform.StackMetadata) (string, error) {
+	buildStackID := os.Getenv(cmd.EnvStackID)
+	if buildStackID == "" {
+		buildStackID = stackMD.BuildImage.StackID
+	}
+
+	if buildStackID == "" {
+		return "", errors.New("CNB_STACK_ID is required when there is no stack metadata available")
+	}
+	return buildStackID, nil
```

Comment:
I found this code a bit confusing. What do you think about:

```suggestion
	var buildStackID string

	if buildStackID = os.Getenv(cmd.EnvStackID); buildStackID != "" {
		return buildStackID, nil
	}
	if buildStackID = stackMD.BuildImage.StackID; buildStackID != "" {
		return buildStackID, nil
	}
	return  "", errors.New("CNB_STACK_ID is required when there is no stack metadata available")
```

or if we don't care about calling `os.Getenv` twice, we can change it to:

```suggestion
	if os.Getenv(cmd.EnvStackID) != "" {
		return os.Getenv(cmd.EnvStackID), nil
	}
	if stackMD.BuildImage.StackID != "" {
		return stackMD.BuildImage.StackID, nil
	}
	return  "", errors.New("CNB_STACK_ID is required when there is no stack metadata available")

```